[Kashish Varshney] iP

kxshxsh opened the pull request:
# Duke
 
 > “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Duke frees your mind of having to remember things you need to do. It's:
- text-based
- easy to learn
- FAST SUPER FAST to use

## Directions:
All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it's **FREE!**

## Some Features:
- [X] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

 If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kxshxsh committed:
Level 1: initial structure + greet, echo, exit

kxshxsh committed:
Implementing adding to and reading the tasklist

kxshxsh committed:
Task class created and mark/unmark feature added to chatbot

kxshxsh committed:
Inheritance used to create 3 types of tasks: todo, deadline and event

kxshxsh committed:
Automated Text UI Testing

kxshxsh committed:
Implemented custom exceptions to handle errors

kxshxsh committed:
Added the delete feature

YiHern-Lee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+                System.out.println((i + 1) + ". " + manager.get(i).toString());
+            }
+        } else {
+            throw new CustomException("your task list is currently empty. " +
```
Perhaps you could break before the "+" operator, to be in accordance with the module's [Java Coding Standards](https://se-education.org/guides/conventions/java/intermediate.html).

YiHern-Lee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+                } else if (instruct.equals("list")) {
+                    reportList();
+                } else {
+                    String[] details = instruct.split(" ");
```
Maybe it will be better if you could break up this chunk of code into different methods? For example, having a method that will take in a `String userInput` as a parameter and then execute the corresponding command.

YiHern-Lee reviewed (commented):
Your code is neat and there is of good quality! However, don't forget to add in Javadocs to help readers better understand your code. 👍

howardwhw2 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+                    reportList();
+                } else {
+                    String[] details = instruct.split(" ");
+                    if (details[0].equals("mark")) {
+                        markAsDone(Integer.parseInt(details[1]));
+                    } else if (details[0].equals("unmark")) {
+                        markNotDone(Integer.parseInt(details[1]));
+                    } else if (details[0].equals("delete")) {
+                        deleteTask(Integer.parseInt(details[1]));
+                    } else {
+                        String taskType = details[0];
+                        if (!(taskType.equals("todo") || taskType.equals("deadline")
+                                || taskType.equals("event"))) {
+                            throw new CustomException("sorry, this isn't a valid command yet!");
+                        } else {
+                            addTask(taskType, instruct);
+                        }
```
Maybe using a switch case for the different commands might make the code easier to read?

howardwhw2 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+        Task desiredTask = new Task("empty task");
+        int len = manager.size();
+
+        if ((num <= len) && (len > 0) && (num > 0)) {
```
perhaps naming your variable names as ```taskNumber``` and ```listLength``` will make it easier to understand?

howardwhw2 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+        int len = manager.size();
+
+        if ((num <= len) && (len > 0) && (num > 0)) {
+            for (int i = 1; i <= len; i++) {
+                if (i == num) {
+                    desiredTask = manager.get(i - 1);
+                }
+            }
```
Maybe you could just change this code to just
```
desiredTask = manager.get(i - 1);
```
which would make the code cleaner

howardwhw2 reviewed (commented):
code quality is good overall but you might want to add some JavaDocs to make your code easier to understand

kxshxsh committed:
save to file, add oop

kxshxsh committed:
add data to gitignore

kxshxsh committed:
adding date and time -- partial

kxshxsh committed:
implement date and time

kxshxsh committed:
Merge branch 'branch-Level-7'

kxshxsh committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

kxshxsh committed:
Improve code's ability to interpret date and times from tasks.txt

s7manth commented on .gitignore:
```diff
@@ -13,5 +13,8 @@ src/main/resources/docs/
+
 /text-ui-test/ACTUAL.txt
 text-ui-test/EXPECTED-UNIX.TXT
+data/
```
`/data/` and `data/` behave differently in `.gitignore` files. 

You might risk ignoring any `data` folders in the entire project if you use `data/`. 

Only `/data/` is enough for our case, since we are ignoring the folder only in the present directory, and not another instance of `data` like `SUBFOLDER/data`.

s7manth commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+
+    public Deadline(String description, String time) {
+        super(description);
+        this.time = LocalDateTime.parse(time, inputFormatter);
```
It would have been better if Deadline could handle the `InvalidParseException` which would occur when the date entered is in the incorrect format. 

By handling this exception, you can prompt the user to use the correct format in the message.

s7manth commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,28 @@
+        super(description);
+        this.dateAndStartTime = LocalDateTime.parse(dateAndStartTime, inputFormatterStart);
+        this.endTime = LocalTime.parse(endTime, inputFormatterEnd);
+    }
```
Similar to the Deadline case, could have handled incorrect parses.

s7manth commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,75 @@
+            if (!(f.exists() && !f.isDirectory())) {
+                f.getParentFile().mkdirs();
+                f.createNewFile();
+            }
```
Great handling the case where the directory itself is missing!

s7manth commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    protected static ArrayList<Task> MANAGER;
```
There is no need for the `MANAGER` to be capital. Only final constants are supposed to be all capital.

s7manth commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+                toAdd = new Event(description, dateAndStartTime, endTime);
+            } else {
+                throw new CustomException("Incorrect time format: ensure to prefix time with '/at'");
+            }
```
You could have used switch statements and component functions to make the code more modular.

s7manth reviewed (commented):
Overall the project looks good. Some suggestions include, handling a more broader range of exceptions, making the code more modular and resistant to future changes, and adding Java Docs for explaining things in a better manner to a human reader.

kxshxsh committed:
code implemented to follow OOP more closely

kxshxsh committed:
further abstracted out functions into the parser class

kxshxsh committed:
packages all classes in a package called duke

kxshxsh committed:
corrected packaging

kxshxsh committed:
Implements JUnit to test important features of the code

kxshxsh committed:
Merge branch 'add-gradle-support'

kxshxsh committed:
Update jar main class pointer

kxshxsh committed:
add find feature

kxshxsh committed:
improves code readibility for find feature

kxshxsh committed:
alter build.gradle to support javafx inorder to implement a GUI

kxshxsh committed:
add JavaDocs to Task and all it's subclasses

kxshxsh committed:
add JavaDocs to Duke.java and Storage.java

kxshxsh committed:
add JavaDocs to Parser.java

kxshxsh committed:
finalise JavaDocs

kxshxsh committed:
add switch cases to improve coding standard

kxshxsh committed:
make minimal changes to the find feature

kxshxsh committed:
add assertions in Storage.java to ensure tasks.txt file creation

kxshxsh committed:
improve code quality by using switch cases, stronger SLAP and more prominent happy paths

kxshxsh committed:
Merge pull request #2 from kxshxsh/branch-A-CodeQuality

improve code quality

kxshxsh committed:
implement snooze feature

kxshxsh committed:
Set theme jekyll-theme-hacker

kxshxsh committed:
Update README.md

kxshxsh committed:
Merge pull request #3 from kxshxsh/kxshxsh-userguide

Update README.md

kxshxsh committed:
Update README.md

kxshxsh committed:
Update README.md

kxshxsh committed:
rectify bugs in snooze

kxshxsh committed:
rectify bugs in snooze

local variable newString in method snooze of class TaskList is
initialised to an empty string.

works well when criteria for snooze is "both" date and time. However, in
the case of "time" only, the newTime variable does not get appropriately
updated leading to parsing errors.

intialising newTime directly to information[2] so that if the change
criteria is "time" only then an empty string will not be parsed.
If change criteria is "both" then newTime will be appropriately
updated to data[1].

Since the steps of updating date and time overlap with updating time
only, it is more logical to, conditionally, update date first. In such
an event, the date is separated from the input and then passed into
newTime for time updation. In the event that date is not meant to be
updated, the newTime string remains empty and leads to errors parsing.
Therefore, if it is directly initialised to the needed argument for
setTime() then errors can be evaded.

kxshxsh committed:
Update README.md

kxshxsh committed:
Merge branch 'master' of https://github.com/kxshxsh/ip

kxshxsh committed:
add minimal duke gui using javafx

kxshxsh committed:
add necessary java fx files

kxshxsh committed:
modify profile pictures for user and duke

uses standard, shared DaDuke and DaUser png files for profile pictures

in order to personalise the chat window and Duke's representation

New non-copyrighted images are cropped in a 1:1 ratio and stored in resources/images under the same names of DaDuke.png and DaUser.png.

Preserving the original file names helps reduce the changes required to be made in the code. Thus reducing unforseen errors.

Image credits:
DaUser: https://unsplash.com/photos/AwCPMLL-fPc
DaDuke: https://unsplash.com/photos/Yv2HzOr8jAA

kxshxsh committed:
Merge pull request #4 from kxshxsh/branch-Level-10

Branch level 10: GUI

kxshxsh committed:
add Ui.png 

adding a representative image of the program GUI
